lib/src/components/AbstractDialog/AbstractDialog.tsx

```diff
@@ -62,108 +79,124 @@ const AbstractDialog = forwardRef<HTMLDialogElement, Props>((props, ref) => {
+      if (isModal) {
+        const autofocusEl = dialogEl.querySelector('[autofocus]') as HTMLElement | null;
+
+        if (autofocusEl) {
+          autofocusEl.focus();
         }
       }
       else {
-        // without native support we don't trap focus in the dialog, but we can at least start it off there
-        // eslint-disable-next-line @typescript-eslint/no-explicit-any
-        (el as any).open = true;
-        el.focus();
+        const dialogFocusableElements = getFirstVisibleFocusableElement(dialogEl);
+        if (dialogFocusableElements) {
+          dialogFocusableElements.focus();
+        }
       }
 
+      // Focus on previously focused element.
       return () => {
-        if (previouslyFocusedEl && previouslyFocusedEl instanceof HTMLElement) {
+        if (previouslyFocusedElementRef.current && previouslyFocusedElementRef.current instanceof HTMLElement) {
           // The useEffect cleanup executes while the dialog is still present (in React 16 at least). While the dialog
           // still exists, the document is still "blocked by the modal dialog" so trying to focus elements outside of
           // it won't work. So we have to wait until the next cycle of the event loop when it's gone
-          Promise.resolve().then(() => previouslyFocusedEl.focus());
+          Promise.resolve().then(() => {
+            if (previouslyFocusedElementRef.current) {
+              previouslyFocusedElementRef.current.focus();
+              previouslyFocusedElementRef.current = null;
+            }
+          });
         }
       };
     }
+    // Close Dialog
     else if (open === false) {
-      if (hasNativeModalSupport) {
+      if (hasNativeModalSupport) { // Use dialog method
         // eslint-disable-next-line @typescript-eslint/no-explicit-any
-        (el as any).close();
+        (dialogEl as any).close();
       }
-      else {
+      else { // Use open attribute
         // eslint-disable-next-line @typescript-eslint/no-explicit-any
-        (el as any).open = false;
+        (dialogEl as any).open = false;
       }
       return undefined;
     }
     else {
       return undefined;
     }
-  }, [open]);
+  }, [open, isModal]);
+
+  // Focus trapping
+  useEffect(() => {
+    const keydownListener = (event: KeyboardEvent) => {
+    // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
+      const dialogEl = dialogRef.current!;
 
-  // listen to the native HTMLDialogElement cancel event which supporting browsers fire when the dialog is closed
-  // via ESC
+      if (event.key === 'Tab') {
+        const firstFocusableElement = getFirstVisibleFocusableElement(dialogEl);
+        const lastFocusableElement = getLastVisibleFocusableElement(dialogEl);
+        const nullOrDialogActiveFocus = [document.body, dialogEl, null].includes(document.activeElement as HTMLElement);
+
+        if (nullOrDialogActiveFocus) {
+          event.preventDefault();
+        }
+
+        if (event.shiftKey) {
```

Comment:
Maybe we should add a few comments to these conditions?

Reply:
> Maybe we should add a few comments to these conditions?

Added comment here: b772c8b9e314502c46bc7a9adbe27e880f96ffee